vktrace/vktrace_common/vktrace_filelike.c

```diff
@@ -88,14 +88,57 @@ uint64_t vktrace_FileLike_GetFileLength(FILE* fp) {
 }
 
 // ------------------------------------------------------------------------------------------------
-FileLike* vktrace_FileLike_create_file(FILE* fp) {
+FileLike* vktrace_FileLike_create_file(FILE* fp, BOOL preload) {
     FileLike* pFile = NULL;
     if (fp != NULL) {
         pFile = VKTRACE_NEW(FileLike);
-        pFile->mMode = File;
-        pFile->mFile = fp;
-        pFile->mMessageStream = NULL;
+        pFile->mFile = NULL;
         pFile->mFileLen = vktrace_FileLike_GetFileLength(fp);
+        pFile->mMessageStream = NULL;
+        pFile->mMemAddr = NULL;
+        pFile->mMemCurrAddr = NULL;
+
+        if (pFile->mFileLen == 0) {
+            vktrace_LogError("Failed to create FileLike, file length is 0!");
+            return pFile;
+        }
+
+        if (preload) {
+            vktrace_LogDebug("Preload trace file to memory.");
+            pFile->mMode = Memory;
+            char* addr = NULL;
+
+#if !defined(WIN32)
+            // mmap does not work on Windows
+            addr = (char*)mmap(NULL, pFile->mFileLen, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_POPULATE, fileno(fp), 0);
+            if (addr == MAP_FAILED) {
+                vktrace_LogError("Failed to mmap tracefile for replaying.");
+                addr = NULL;
+            }
+#else
+            // TODO: Find a better solution to make it work with a large trace file
+            addr = VKTRACE_NEW_ARRAY(char, pFile->mFileLen);
+            if (addr == NULL) {
+                vktrace_LogError("Failed to malloc memory to load tracefile for replaying.");
+            } else {
+                if (1 != fread(addr, (size_t)pFile->mFileLen, 1, fp)) {
+                    if (ferror(fp) != 0) {
+                        perror("fread error");
+                    } else if (feof(fp) != 0) {
+                        vktrace_LogVerbose("Reached end of file.");
```

Comment:
Why would this condition occur?

Reply:
It was added to handle empty file, but yes, the 0 file length is already handled by line 101 "if (pFile->mFileLen == 0)", I will remove this condition check.